features/topology-map/plugins/org.opennms.features.topology.plugins.topo.pathoutage/src/main/java/org/opennms/features/topology/plugins/topo/pathoutage/PathOutageProvider.java

```diff
@@ -0,0 +1,184 @@
+ * by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * OpenNMS(R) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with OpenNMS(R).  If not, see:
+ *      http://www.gnu.org/licenses/
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.features.topology.plugins.topo.pathoutage;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.opennms.features.topology.api.browsers.ContentType;
+import org.opennms.features.topology.api.browsers.SelectionChangedListener;
+import org.opennms.features.topology.api.support.VertexHopGraphProvider;
+import org.opennms.features.topology.api.topo.AbstractEdge;
+import org.opennms.features.topology.api.topo.AbstractTopologyProvider;
+import org.opennms.features.topology.api.topo.AbstractVertex;
+import org.opennms.features.topology.api.topo.Defaults;
+import org.opennms.features.topology.api.topo.Status;
+import org.opennms.features.topology.api.topo.Vertex;
+import org.opennms.features.topology.api.topo.VertexRef;
+import org.opennms.netmgt.dao.api.NodeDao;
+import org.opennms.netmgt.model.OnmsNode;
+import org.opennms.netmgt.model.OnmsSeverity;
+
+import com.google.common.collect.Lists;
+
+import edu.uci.ics.jung.graph.DirectedSparseGraph;
+import edu.uci.ics.jung.graph.util.EdgeType;
+
+/**
+ * This provider allows us to build a vertex hierarchy, based on parent-child relations between nodes
+ */
+public class PathOutageProvider extends AbstractTopologyProvider {
+
+	public static final String NAMESPACE = "pathoutage";
+
+	private NodeDao nodeDao;
+	private PathOutageStatusProvider statusProvider;
+	private DirectedSparseGraph<PathOutageVertex, AbstractEdge> sparseGraph;
+
+	public PathOutageProvider(NodeDao nodeDao, PathOutageStatusProvider pathOutageStatusProvider) {
+		super(NAMESPACE);
+		this.nodeDao = nodeDao;
+		this.statusProvider = pathOutageStatusProvider;
+		this.sparseGraph = null;
+	}
+
+	@Override
+	public void refresh() {
+		resetContainer();
+		load();
+	}
+
+	@Override
+	public Defaults getDefaults() {
+		// Current implementation of default focus strategy includes all "problematic" (state not NORMAL) nodes, semantic zoom level is 0
+		return new Defaults()
+				.withSemanticZoomLevel(0)
+				.withPreferredLayout("Hierarchy Layout")
+				.withCriteria(() -> {
+					OnmsSeverity worstSeverity = OnmsSeverity.NORMAL;
```

Comment:
I'd prefer to use the `getStatusForVertices(...) ` method for all vertices to reduce the database queries.
The focus calculation may look like this:
```java
Optional<Map.Entry<VertexRef, Status>> max = resultMap.entrySet().stream().max(Comparator.comparing(e -> OnmsSeverity.get(e.getValue().computeStatus())));
        if (max.isPresent()) {
            focusVertex = max.get().getKey();
        } else if(vertexProvider.getVertexTotalCount() > 0) {
            focusVertex = vertexProvider.getVertices().get(0);
        } else {
          focusVertex = null;
       }
```